src/cudamatrix/cu-kernels.cu

```diff
@@ -1763,6 +1763,31 @@ static void _group_max(Real *y, const Real *x, MatrixDim d, int src_stride,
   }
 }
 
+template<typename Real>
+__global__
+static void _parametric_relu(Real*y, const Real*x, MatrixDim d, int src_stride, const Real*a, const Real*b) {
+  int i = blockIdx.x * blockDim.x + threadIdx.x;
+  int j = blockIdx.y * blockDim.y + threadIdx.y;
+  int dst_index = i + j*d.stride, src_index = i + j * src_stride;
+  if(i < d.cols && j < d.rows) {
+    Real res = (x[src_index] > 0.0) ? a[i] * x[src_index] : b[i] * x[src_index];
+    y[dst_index] = res;
+  }
+}
+
+
+template<typename Real>
+__global__
+static void _diff_parametric_relu(Real*eout, const Real*e, const Real*y, MatrixDim d, int e_stride, int y_stride, const Real*a, const Real*b) {
+  int i = blockIdx.x * blockDim.x + threadIdx.x;
+  int j = blockIdx.y * blockDim.y + threadIdx.y;
+  int dst_index = i + j*d.stride;
+  int e_index   = i + j*e_stride;
+  int y_index   = i + j*y_stride;
+  if (i < d.cols  && j < d.rows )
+    eout[dst_index] = (y[y_index] > 0.0 ? a[i] * e[e_index] : b[i] * e[e_index]);
```

Comment:
Aha, I see. I would also expect that `x` is corresponding to `in` and `y` to `out`. I was thinking there could be a possible situation in which `alpha` or `beta` becomes negative. Then it would be crucial to use `in` -> `x`.

But still ... , If the `PropagateFnc` is doing the decision based on `x` it feels natural that the decision in `BackpropagateFnc` is also based on `x`.